java-frontend/src/main/java/org/sonar/java/caching/ContentHashCache.java

```diff
@@ -0,0 +1,117 @@
+import org.sonar.api.batch.sensor.cache.ReadCache;
+import org.sonar.api.batch.sensor.cache.WriteCache;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+import java.io.IOException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+public class ContentHashCache {
+
+  private static final Logger LOG = Loggers.get(ContentHashCache.class);
+  private static final String CONTENT_HASH_KEY = String.format("java:contentHash:%s:", FileHashingUtils.HASH_ALGORITHM);
+  private static final String HASH_COMPUTE_FAIL_MSG = "Failed to compute content hash for file %s";
+
+  private ReadCache readCache;
+  private WriteCache writeCache;
+  private final boolean enabled;
+
+  public ContentHashCache(SensorContext context) {
+    CacheContextImpl cacheContext = CacheContextImpl.of(context);
+    enabled = cacheContext.isCacheEnabled();
+
+    if (enabled) {
+      readCache = context.previousCache();
+      writeCache = context.nextCache();
+    }
+  }
+
+  public boolean hasSameHashCached(InputFile inputFile) {
+    if (!enabled) {
+      LOG.trace("Cannot read from cache when disabled");
+      return false;
+    }
+    String cacheKey = getCacheKey(inputFile);
+    try {
+      LOG.trace("Reading cache for the file {}", inputFile.key());
+      byte[] cachedHash = readCache.read(cacheKey).readAllBytes();
+      byte[] fileHash = FileHashingUtils.inputFileContentHash(inputFile);
+      boolean isHashEqual = MessageDigest.isEqual(fileHash, cachedHash);
+      if (isHashEqual) {
+        copyFromPrevious(inputFile);
+      } else {
+        writeToCache(inputFile);
+      }
+      return isHashEqual;
+    } catch (IllegalArgumentException e) {
+      LOG.trace(String.format("Could not find key %s in the cache", cacheKey));
+      writeToCache(inputFile);
+    } catch (IOException | NoSuchAlgorithmException e) {
+      LOG.trace(String.format(HASH_COMPUTE_FAIL_MSG, inputFile.key()));
+    }
+    return false;
+  }
+
+  public boolean contains(InputFile inputFile) {
+    if (!enabled) {
+      LOG.trace("Cannot read from cache when disabled");
+      return false;
+    }
+    return readCache.contains(getCacheKey(inputFile));
+  }
+
+  public boolean writeToCache(InputFile inputFile) {
+    if (!enabled) {
+      LOG.trace("Cannot write on cache when disabled");
+      return false;
+    }
+    if (writeCache != null) {
+      LOG.trace(String.format("Writing to the cache for file %s", inputFile.key()));
+      String cacheKey = getCacheKey(inputFile);
+      try {
+        writeCache.write(cacheKey, FileHashingUtils.inputFileContentHash(inputFile));
+        return true;
+      } catch (IllegalArgumentException e) {
+        LOG.error(String.format("Tried to write multiple times to cache key '%s'. Ignoring writes after the first.", cacheKey));
+      } catch (IOException | NoSuchAlgorithmException e) {
+        LOG.error(String.format(HASH_COMPUTE_FAIL_MSG, inputFile.key()));
```

Comment:
We should change the level of these logs to warning because while it may be interesting to investigate for us, failure to write to the cache should not affect analysis correctness